pmlc/dialect/pxa/transforms/reorder_layout.cc

```diff
@@ -0,0 +1,264 @@
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+
+class ReorderLayoutPass final : public ReorderLayoutBase<ReorderLayoutPass> {
+public:
+  void runOnFunction();
+};
+
+struct ReorderDesc {
+  mlir::AffineMap reorderMap;
+  mlir::SmallVector<uint64_t, 6> reorderedShape;
+  mlir::SmallVector<uint64_t, 6> reorderedVector;
+};
+
+/// Simplify affine map given integral constraints. Returns the same map if it
+/// cannot be simplified further.
+///
+/// Input:
+///   map         = (d0, d1) -> (d0 + d1, (d0 * 16 + d1 * 8) floordiv 8 floordiv
+///   2,
+///                              (d0 * 16 + d1 * 8) floordiv 8 % 2, 0)
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+///
+/// Output:
+///   (d0, d1) -> (d0 + d1, d0, d1, 0)
+mlir::AffineMap simplifyConstrainedAffineMap(mlir::AffineMap map,
+                                             mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  return map;
+}
+
+/// Expand affine map dimensions based on integral constraints and vector shape.
+///
+/// Input:
+///   map (A)     = (d0, d1) -> (d0 + d1, d0 * 16 + d1 * 8)
+///   shape       = <7, 96>
+///   vector      = <1, 8>
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+/// Output:
+///   reorder map (B)  = (d0, d1) -> (d0, d1 floordiv 8 floordiv 2, d1 floordiv
+///   8 % 2, 0) reordered shape  = <7, 6, 2, 8> reordered vector = <1, 1, 1, 8>
+///
+/// Note: to obtain affine map from input space to expanded space composition
+///       A o B can be used (with simplification).
+/// A o B = (d0, d1) -> (d0 + d1, d0, d1, 0)
+ReorderDesc expandAffineMap(mlir::AffineMap map, mlir::ArrayRef<uint64_t> shape,
+                            mlir::ArrayRef<uint64_t> vector,
+                            mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  auto reorderMap = mlir::AffineMap::getMultiDimIdentityMap(map.getNumResults(),
+                                                            map.getContext());
+  mlir::SmallVector<uint64_t, 6> reorderedShape(shape.begin(), shape.end());
+  mlir::SmallVector<uint64_t, 6> reorderedVector(vector.begin(), vector.end());
+  return ReorderDesc{reorderMap, reorderedShape, reorderedVector};
+}
+
+/// Create affine permutation map that sorts resulting space dimensions in order
+/// of increasing schedule.
+///
+/// Input:
+///   map (A)    = (d0, d1) -> (d0 + d1, d0, d1, 0)
+///   shape      = <7, 6, 2, 8>
+///   vector     = <1, 1, 1, 8>
+///   schedule   = <1, 0>
+/// Output:
+///   reorder map (B)  = (d0, d1, d2, d3) -> (d2, d0, d1, d3)
+///   reordered shape  = <2, 7, 6, 8>
+///   reordered vector = <1, 1, 1, 8>
+///
+/// Note: to obtain affine map from input space to sorted space composition
+///       A o B can be used.
+///   A o B = (d0, d1) -> (d1, d0 + d1, d0)
+ReorderDesc sortAffineMap(mlir::AffineMap map, mlir::ArrayRef<uint64_t> shape,
+                          mlir::ArrayRef<uint64_t> vector,
```

Comment:
Also a bit unclear in intent. Seems to be doing a permutation where the dimensions in the 'split' version of the tensor are sorted by whether or not the use a certain input dimension (i.e. dim 0 and 2 seems to be moved to the front, presumable since they used d1, which is 0 in the schedule?). Also, the composition doesn't make sense since it should have a 4d output I think?

Reply:
Yes, it should be 4d, I missed that.
I have added some loose set of rules that hint at order for which we want to sort, but I'm not really sure how to convey this most clearly without explicitly defining strict weak ordering here.
I feel that best way to think about this is in reverse order, since more frequently changing dimension appears later in affine schedule, so they should also appear later in map.
So to sort for given schedule in basic case we are interested in dimension that appears latest in schedule.

In this case, d3 (0) is last as that's vector dimension, next is d1 (A: d0) as it uses only d0 in input which is last in schedule.
Then d0 (A: d0 + d1), because it uses d0, but addition of d1 makes it before only d0. Finally d2 (A: d1) as previous in schedule.

Reply:
Also worth to note that what sorting aims to do is to order dimensions in terms of how often they will repeat same value (least frequently to most), but I have no idea how to measure this exactly with non-trivial maps, hence this approximation.